omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java

```diff
@@ -120,33 +114,58 @@ public String get(PageModel model,
 		return null;
 	}
 
-	private String getRedirectUrlFromReferer(PageRequest pageRequest) {
-	    String referer = pageRequest.getRequest().getHeader("Referer");
-	    String redirectUrl = "";
-		if (referer != null) {
-			String manualLogout = pageRequest.getSession().getAttribute(AppUiConstants.SESSION_ATTRIBUTE_MANUAL_LOGOUT, String.class);
-			if (!"true".equals(manualLogout)) {
-				if (referer.contains("http://") || referer.contains("https://")) {
-					try {
-						URL refererUrl = new URL(referer);
-						String refererPath = refererUrl.getFile();
-						String refererContextPath = refererPath.substring(0, refererPath.indexOf('/', 1));
-						if (StringUtils.equals(pageRequest.getRequest().getContextPath(), refererContextPath)) {
-							redirectUrl = refererPath;
-						}
-					}
-					catch (MalformedURLException e) {
-						log.error(e.getMessage());
-					}
-				} else {
-					redirectUrl = pageRequest.getRequest().getHeader("Referer");
-				}
+	private boolean isUrlWithinOpenmrs(PageRequest pageRequest, String redirectUrl){
+		if (StringUtils.isNotBlank(redirectUrl)) {
+			if (redirectUrl.startsWith("http://") || redirectUrl.startsWith("https://")) {
+				try {
+                    URL url = new URL(redirectUrl);
+                    String urlPath = url.getFile();
+                    String urlContextPath = urlPath.substring(0, urlPath.indexOf('/', 1));
+                    if (StringUtils.equals(pageRequest.getRequest().getContextPath(), urlContextPath)) {
+                        return true;
+                    }
+                }
+                catch (MalformedURLException e) {
+                    log.error(e.getMessage());
+                }
+			} else if(redirectUrl.startsWith("/openmrs")){
```

Comment:
What happens when one is using a webapp context name other than openmrs? Something like bahmni?